designs/resourceCache/design.md

```diff
@@ -0,0 +1,195 @@
+# AWS Resource Cache
+
+List and describe calls to AWS services are used to fetch information about the resources 
+available within a customers account. Typical examples are:
+ - populate AWS explorer views
+ - populate resource-selection drop-down boxes
+
+In some cases the lifecycle of these describe calls is not controlled by the plugin but rather
+the [IntelliJ SDK][1] - we don't control when this data is refreshed. An example is the [`LambdaLineMarker`][2] which implements [`LineMarkerProvider`][3],
+the [`LineMarkerProvider.getLineMarkerInfo`][4]
+is invoked repeatedly as a file is being edited to determine if a line marker should be shown.
+
+AWS resources are rarely provisioned multiple times per second and network bandwidth on developer laptops as
+well as AWS throttling limits makes repeatedly making list/describe calls for resources untenable. Clearly a cache
+is required.
+
+## Requirements
+
+The resource cache has some unique requirements that make sense to implement in a re-usable way on-top of a
+known in-memory caching implementation ([Guava][5]):
+- Ability to cache a resource (or list of resources) by AWS [Region][6] / Account ([Credential][7])
+- Time-based expiration, customizable by resource-type (some resources are more dynamic than others)
+- The ability to fall-back to stale data in the event of a failed refresh (e.g. the developers machine is offline)
+- Ability to force a refresh when we know a resource has changed
+- Be respectful of a developer machines memory constraints
+
+## Implementation
+The following section describes the implementation of the [`AwsResourceCache`][8] 
+inside the toolkit. The [`AwsResourceCache`][8] is a `Project` level service that exposes methods for fetching a 
+resource (either from the cache or by making a network call), as well as methods for clearing existing cached resources
+on a number of dimensions.
+
+The cache also understands when credential providers have been mutated (by implementing `ToolkitCredentialsChangeListener`) this
```

Comment:
what if the customer changes profiles frequently? should we throw the data away when they switch profiles? or just deprioritize it? i.e. move it out of the current scope.  could we put the "out of scope data" first in the line in case of GC?